File: WeekEndRayTrace/src/Vec3.java

public class Vec3 {
	public final double x;
	public final double y;
	public final double z;

	public Vec3(double x, double y, double z) {
		this.x = x;
		this.y = y;
		this.z = z;
	}

	public double length() {
		return Math.sqrt(squaredLength());
	}

	public double squaredLength() {
		return x * x + y * y + z * z;
	}

	public Vec3 add(Vec3 v) {
		return new Vec3(x + v.x, y + v.y, z + v.z);
	}

	public Vec3 mul(double s) {
		return new Vec3(x * s, y * s, z * s);
	}

	public Vec3 div(double s) {
		return new Vec3(x / s, y / s, z / s);
	}

	public Vec3 unit() {
		return div(length());
	}

}

File: KotRay/src/kotray/Vec3.kt
package kotray

import  java.lang.Math.sqrt

/**
 * 3次元ベクトル
 *
 * @param x :
 * @param y :
 * @param z :
 */
data class Vec3(val x: Double, val y: Double, val z: Double) {
    /** @return ベクトルの長さ */
    val length: Double
        get() = sqrt(squaredLength)

    /** @return ベクトルの長さの二乗 */
    val squaredLength: Double
        get() = x * x + y * y + z * z

    /** @return 長さを1にしたベクトル */
    val unit: Vec3
        get() = div(length)

    operator fun unaryMinus(): Vec3 = Vec3(-x, -y, -z)

    operator fun plus(v: Vec3): Vec3 = Vec3(x + v.x, y + v.y, z + v.z)

    operator fun minus(v: Vec3): Vec3 = Vec3(x - v.x, y - v.y, z - v.z)

    operator fun times(s: Double): Vec3 = Vec3(x * s, y * s, z * s)

    operator fun times(v: Vec3): Vec3 = Vec3(x * v.x, y * v.y, z * v.z)

    operator fun div(s: Double): Vec3 = Vec3(x / s, y / s, z / s)
}

/**
 * 内積
 */
fun dot(v1: Vec3, v2: Vec3): Double = v1.x * v2.x + v1.y * v2.y + v1.z * v2.z

/**
 * 外積
 */
fun cross(v1: Vec3, v2: Vec3): Vec3 {
    return Vec3(
            v1.y * v2.z - v1.z * v2.y,
            v1.z * v2.x - v1.x * v2.z,
            v1.x * v2.y - v1.y * v2.x)
}
